packages/@uppy/companion-client/src/RequestClient.js

```diff
@@ -94,92 +89,82 @@ export default class RequestClient {
     return `${this.hostname}/${url}`
   }
 
-  #errorHandler (method, path) {
-    return (err) => {
-      if (!err?.isAuthError) {
-        // eslint-disable-next-line no-param-reassign
-        err = new ErrorWithCause(`Could not ${method} ${this.#getUrl(path)}`, { cause: err })
+  /*
+    Preflight was added to avoid breaking change between older Companion-client versions and
+    newer Companion versions and vice-versa. Usually the break will manifest via CORS errors because a
+    version of companion-client could be sending certain headers to a version of Companion server that
+    does not support those headers. In which case, the default preflight would lead to CORS.
+    So to avoid those errors, we do preflight ourselves, to see what headers the Companion server
+    we are communicating with allows. And based on that, companion-client knows what headers to
+    send and what headers to not send.
+
+    The preflight only happens once throughout the life-cycle of a certain
+    Companion-client <-> Companion-server pair (allowedHeadersCache).
+    Subsequent requests use the cached result of the preflight.
+    However if there is an error retrieving the allowed headers, we will try again next time
+  */
+  async preflight (path) {
+    const allowedHeadersCached = allowedHeadersCache.get(this.hostname)
+    if (allowedHeadersCached != null) return allowedHeadersCached
+
+    const promise = (async () => {
+      try {
+        const response = await fetch(this.#getUrl(path), { method: 'OPTIONS' })
+
+        const header = response.headers.get('access-control-allow-headers')
+        if (header == null) return undefined
+
+        this.uppy.log(`[CompanionClient] adding allowed preflight headers to companion cache: ${this.hostname} ${header}`)
+
+        return header.split(',').map((headerName) => headerName.trim().toLowerCase())
+      } catch (err) {
+        this.uppy.log(`[CompanionClient] unable to make preflight request ${err}`, 'warning')
```

Comment:
```suggestion
        this.uppy.log(`[CompanionClient] unable to make preflight request ${err}`, 'warning')
        allowedHeadersCache.delete(this.hostname)
```

Reply:
I thought if companion OPTIONS responds without any access-control-allow-headers, that's also a bug and we should not cache that, but I'm not sure about that..

Reply:
I think that's when we are receiving a response from an older Companion instance.